server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -310,6 +311,49 @@ public void parse(ParseContext context) throws IOException {
      */
     protected abstract void parseCreateField(ParseContext context) throws IOException;
 
+    /**
+     * Given access to a document's _source, return this field's values.
+     *
+     * In addition to pulling out the values, mappers can parse them into a standard form. This
+     * method delegates parsing to {@link #parseSourceValue} for parsing. Most mappers will choose
+     * to override {@link #parseSourceValue} -- for example numeric field mappers make sure to
+     * parse the  source value into a number of the right type.
+     *
+     * Some mappers may need more flexibility and can override this entire method instead.
+     *
+     * @param lookup a lookup structure over the document's source.
+     * @return a list a standardized field values.
+     */
+    public List<?> lookupValues(SourceLookup lookup) {
+        Object sourceValue = lookup.extractValue(name());
+        if (sourceValue == null) {
+            return List.of();
+        }
+
+        List<Object> values = new ArrayList<>();
+        if (this instanceof ArrayValueMapperParser) {
```

Comment:
Is there a way we can avoid the `instanceof` somehow? I feel like we have so so so many of these already and they drive me crazy. They make it so hard to read classes and know what is up.

Reply:
`ArrayValueMapperParser` is a marker interface whose sole purpose to indicate to the parsing logic that `FieldMapper#parse` handles array values, and that they shouldn't be split up. I liked how this logic mirrored the document parsing logic in how it handled array values.

After you mentioned this, I looked into refactoring how we handle array-value parsing in general so we don't rely on `mapper instanceof ArrayValueMapperParser` checks. I wasn't able to refactor it successfully because of a complexity with `copy_to` fields. (In general I've been finding it really tricky to make incremental improvements to the document parsing code!)

Reply:
Bleh! Well, adding this won't make it worse then! Marker interfaces are not my friend! But it isn't like you can do anything about it now.....